### PR TITLE
Fall back to org.gnome.desktop.interface if wingpanel datetime schema does not exist.

### PR DESCRIPTION
**Pull request by JoshStrobl**
This commit will handle the case where the io.elementary.desktop.wingpanel.datetime schema doesn't exist, which would otherwise be fatal on platforms which support / ship applications which use `Granite.DateTime.is_clock_format_12h()` but do not support / ship Wingpanel out-of-the-box.

**Note:** This patch was tested under a non-elementary OS system that does not ship or provide Wingpanel and its respective schemas, against a Granite-based application (Tootle). If this patch looks good to you folks, I'd still strongly encourage that someone from elementary test to validate it is indeed utilizing the `io.elementary.desktop.wingpanel.datetime` schema.

### CODE REVIEW

File: lib/DateTime.vala [hunk 1]
```diff
@@ -110,9 +112,21 @@ namespace Granite.DateTime {
      * @return true if the clock format is 12h based, false otherwise.
      */
     private static bool is_clock_format_12h () {
-        var h24_settings = new Settings ("io.elementary.desktop.wingpanel.datetime");
-        var format = h24_settings.get_string ("clock-format");
-        return (format.contains ("12h"));
+        SettingsSchemaSource? source = SettingsSchemaSource.get_default();
```
Comment (jeremypw) [inline lib/DateTime.vala]: Elementaryos code style: missing space before bracket.

File: lib/DateTime.vala [hunk 2]
```diff
@@ -110,9 +112,21 @@ namespace Granite.DateTime {
...
+        bool use_wingpanel_schema = false;
+
+        if (source != null) { // If we have a valid SettingsSchemaSource
+                use_wingpanel_schema = (source.lookup(WINGPANEL_DATETIME_SCHEMA, true) != null);
```
Comment (jeremypw) [inline lib/DateTime.vala]: Elementaryos code style: missing space before bracket.

File: lib/DateTime.vala [hunk 3]
```diff
@@ -110,9 +112,21 @@ namespace Granite.DateTime {
...
+        }
+
+        Settings h24_settings = new Settings(use_wingpanel_schema ? WINGPANEL_DATETIME_SCHEMA : "org.gnome.desktop.interface");
```
Comment (jeremypw) [inline lib/DateTime.vala]: Elementaryos code style: missing space before bracket.

File: lib/DateTime.vala [hunk 4]
```diff
@@ -110,9 +112,21 @@ namespace Granite.DateTime {
...
+                
+        if (h24_settings != null) { // Neither schema exists
```
Comment (jeremypw) [inline lib/DateTime.vala]: Comment should be against `} else {`

File: lib/DateTime.vala [hunk 5]
```diff
@@ -22,6 +22,8 @@
  * getting the default translated format for either date and time.
  */
 namespace Granite.DateTime {
+    private const string WINGPANEL_DATETIME_SCHEMA = "io.elementary.desktop.wingpanel.datetime";
+
```
Comment (jeremypw) [inline lib/DateTime.vala]: You could make the gnome schema a constant as well.

**Comment by jeremypw** [comment]
I can confirm that, on Juno, the wingpanel and calendar continued to work as normal, and respond to the wingpanel clock-format setting when this PR was installed and the system rebooted.

**Comment by bleakgrey** [comment]
Any chance of merging this PR soon? Tootle depends on Granite's relative date parsing. As a result, Tootle's been crashing on non-elementary OS systems since October.

**Comment by JoshStrobl** [comment]
Yea apologies, I'll be correcting the various issues @jeremypw raised today, just been busy with work.

File: lib/DateTime.vala [hunk 6]
```diff
@@ -110,9 +112,21 @@ namespace Granite.DateTime {
...
+        bool use_wingpanel_schema = false;
+
+        if (source != null) { // If we have a valid SettingsSchemaSource
+                use_wingpanel_schema = (source.lookup(WINGPANEL_DATETIME_SCHEMA, true) != null);
+        }
+
+        Settings h24_settings = new Settings(use_wingpanel_schema ? WINGPANEL_DATETIME_SCHEMA : "org.gnome.desktop.interface");
+                
+        if (h24_settings != null) { // Neither schema exists
+                var format = h24_settings.get_string ("clock-format");
```
[the remark this answers is not in the document]
Comment (JoshStrobl) [reply lib/DateTime.vala]: I'm going to mark this as resolved. There is no safe method to check if the specific key exists, nor do I believe it is really necessary. `clock-format` has been around in the GNOME interface schema for over 8 years, I seriously doubt it is going away anytime soon. The key in the granite schema is something you folks obviously have control over and seeing how it was recently added, I'm also doubtful this is something that's going to be suddenly removed.

**Comment by JoshStrobl** [comment]
Styling issues should now be addressed.

**Comment by worldofpeace** [comment]
https://github.com/elementary/granite/pull/273 is merged now so this won't be needed anymore.

**Pull request closed without merging**